frontend/src/app/shared/transaction.utils.ts

```diff
@@ -2157,3 +2136,283 @@ const opcodes = {
+};
+
+function computeMerkleRoot(scriptHex: string, leafVersion: number, merkleBranches: string[]): string {
+  const leafHash = computeLeafHash(scriptHex, leafVersion);
+  return (merkleBranches || []).reduce((acc, branch) => {
+    const [firstChild, secondChild] = [acc, branch].sort((a, b) => a.localeCompare(b));
+    return taggedHash('TapBranch', firstChild + secondChild);
+  }, leafHash);
+}
+
+function deriveTaprootOutput(internalKey: string, merkleRoot: string, network: string): { address: string; parity: number } {
+  const tweakHash = taggedHash('TapTweak', internalKey + merkleRoot); // HashTapTweak(internalKey || m)
+  const tweak = BigInt('0x' + tweakHash) % SECP256K1_ORDER; // int(HashTapTweak(internalKey || m))
+  const internalKeyPoint = secp256k1.Point.fromHex(`02${internalKey}`); // P = lift_x(internalKey)
+  const outputKeyPoint = internalKeyPoint.add(secp256k1.Point.BASE.multiply(tweak)); // Q = P + int(HashTapTweak(internalKey || m)) * G
+  const parity = Number(outputKeyPoint.y & 1n);
+  const outputKey = outputKeyPoint.x.toString(16).padStart(64, '0');
+  return { address: scriptPubKeyToAddress(`5120${outputKey}`, network).address, parity };
+}
+
+interface TapTreeLeaf {
+  leafVersion: number;
+  scriptHex: string;
+  merkleBranches: string[];
+}
+
+/** Decode a PSBT_OUT_TAP_TREE into its leaves */
+function parseTapTreeRecord(value: Uint8Array): TapTreeLeaf[] {
+  const leaves: TapTreeLeaf[] = [];
+  const stack: { depth: number; hash: string; leaves: TapTreeLeaf[] }[] = [];
+  let offset = 0;
+
+  while (offset < value.length) {
+    const depth = value[offset++];
+    const leafVersion = value[offset++];
+    const [scriptLength, scriptOffset] = readVarInt(value, offset);
+    offset = scriptOffset;
+    const [scriptBytes, nextOffset] = readSlice(value, offset, scriptLength);
+    offset = nextOffset;
+    const scriptHex = uint8ArrayToHexString(scriptBytes);
+    const leaf: TapTreeLeaf = {
+      leafVersion,
+      scriptHex,
+      merkleBranches: [],
+    };
+    leaves.push(leaf);
+    stack.push({ depth, hash: computeLeafHash(scriptHex, leafVersion), leaves: [leaf] });
+
+    while (stack.length >= 2 && stack[stack.length - 1].depth === stack[stack.length - 2].depth) {
+      const right = stack.pop();
+      const left = stack.pop();
+      for (const l of left.leaves) {
+        l.merkleBranches.push(right.hash);
+      }
+      for (const r of right.leaves) {
+        r.merkleBranches.push(left.hash);
+      }
+      const [firstChild, secondChild] = [left.hash, right.hash].sort((a, b) => a.localeCompare(b));
+      stack.push({ depth: left.depth - 1, hash: taggedHash('TapBranch', firstChild + secondChild), leaves: left.leaves.concat(right.leaves) });
+    }
+  }
+
+  return leaves;
+}
+
+/** Populate an address' taptree using a PSBT involving the taproot address */
+export function fillTapTree(addressTypeInfo: AddressTypeInfo, input: string) {
```

Comment:
I'd suggest splitting out the PSBT parsing and the script processing/taptree reconstruction into separate methods.

for the latter, it seems like you only really need the `AddressTypeInfo` object, an internal key, and a list of `TapTreeLeaf`s.

so you could parse those out from a PSBT or query string param or whatever other input formats we want to support separately, and keep the tree reconstruction logic nicely decoupled.